File: hw1/hw1/Parser.cs
using System;

namespace hw1
{
    public class Parser
    {
        private const int WrongArgCount = 1;
        private const int WrongArgFormat = 2;
        private const int WrongOperation = 3;

        public static int TryParseArguments(string[] args, out Operations operation, out int val1, out int val2)
        {
            operation = default;
            val1 = default;
            val2 = default;
            
            if (!CheckArgumentsCount(args))
                return WrongArgCount;
            
            if (!TryParseOperation(args[1], out operation))
                return WrongOperation;
            
            if (!(TryParseValue(args[0], out val1) && TryParseValue(args[2], out val2)))
            {
                return WrongArgFormat;
            }
            
            return 0;
        }
        
        private static bool TryParseValue(string args, out int val)
        {
            if (int.TryParse(args, out val)) 
                return true;
            Console.WriteLine($"Value is not int: {args}");
            return false;
        }

        private static bool TryParseOperation(string arg, out Operations operation)
        {
            var temp = arg switch
            {
                "+" => 0,
                "-" => 1,
                "*" => 2,
                "/" => 3,
                _ => 4
            };
            if (!(temp < 4))
            {
                operation = default;
                Console.WriteLine($"The calculator does not recognize this operation: {arg}");
                return false;
            }

            operation = (Operations) temp;
            return true;
        }

            private static bool CheckArgumentsCount(string[] args)
        {
            if (args.Length == 3) return true;
            Console.WriteLine($"The program requires 3 arguments to work, but the {args.Length} was entered");
            return false;
        }
    }
}

File: HW4/HW4/Parser.fs
module HW4.Parser

open System
    
let checkArgumentCount (args:string[]) =
    args.Length = 3

let tryParseOperation (arg:string) (operation: outref<Operations>) =
    operation <- match arg with
    | "+" -> Operations.Plus
    | "-" -> Operations.Minus
    | "*" -> Operations.Multiply
    | "/" -> Operations.Divide
    | _ -> Operations.Error
    
let tryParseValue (arg:string) (value: outref<int>) =
    if Int32.TryParse(arg, &value) then
        true
    else
        printf $"Value is not int: {arg}"
        false
        
let tryParseArgs (args:string[]) (operation: outref<Operations>) (val1: outref<int>) (val2: outref<int>) =
    tryParseOperation args.[1] &operation
    if tryParseValue args.[0] &val1 = false || tryParseValue args.[2] &val2 = false then
        ErrorCodes.WrongArgFormat
    elif checkArgumentCount args = false then
        printf $"The program requires 3 arguments to work, but the {args.Length} was entered"
        ErrorCodes.WrongArgCount
    elif operation = Operations.Error then
        printf $"The calculator does not recognize this operation: {args.[1]}"
        ErrorCodes.WrongOperation
    else
        ErrorCodes.Correct